llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp

```diff
@@ -315,6 +315,7 @@ Value *InstCombinerImpl::SimplifyDemandedUseBits(Value *V, APInt DemandedMask,
     if (DemandedMask.isSubsetOf(RHSKnown.Zero | LHSKnown.Zero)) {
       Instruction *Or =
           BinaryOperator::CreateOr(I->getOperand(0), I->getOperand(1));
+      cast<PossiblyDisjointInst>(Or)->setIsDisjoint(true);
```

Comment:
I don't think this is quite correct. This is only checking that the demanded bits are disjoint. If there is a conflict in the non-demanded bits, adding the `disjoint` flag is still going to make the result poison.

Reply:
I've added a check that the demanded mask is all ones.